### PR TITLE
chore: improve distruption for underutilization

### PR DESCRIPTION
**Pull request by Luke-Smartnews**
## Background

we're trying to use karpenter in our production environment, the scaling-out feature is working well, but the scaling-in(disruption) is a problem to us. There're are only two options: `WhenUnderutilized | WhenEmpty` ,
- WhenEmpty: basically useless
- WhenUnderutilized: we see frequently node deletion and creation which reduces SLA. (https://github.com/kubernetes-sigs/karpenter/issues/735)

## Proposals

- add an option to check the node utilization first, normally if a node is fully utilized no need for disruption.
- support consolidateAfter for WhenUnderutilized

### CODE REVIEW

File: pkg/apis/crds/karpenter.sh_nodepools.yaml [hunk 1]
```diff
@@ -3,7 +3,8 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.14.0
+    controller-gen.kubebuilder.io/version: v0.8.0
```
Comment (Bryce-Soghigian) [inline pkg/apis/crds/karpenter.sh_nodepools.yaml:6]: please run `make toolchain` to update your controller-gen version

File: pkg/apis/v1beta1/nodeclaim_status.go [hunk 2]
```diff
@@ -58,12 +58,13 @@ var LivingConditions = []apis.ConditionType{
...
-	Registered  apis.ConditionType = "Registered"
-	Initialized apis.ConditionType = "Initialized"
-	Empty       apis.ConditionType = "Empty"
-	Drifted     apis.ConditionType = "Drifted"
-	Expired     apis.ConditionType = "Expired"
+	Launched     apis.ConditionType = "Launched"
+	Registered   apis.ConditionType = "Registered"
+	Initialized  apis.ConditionType = "Initialized"
+	Empty        apis.ConditionType = "Empty"
+	Consolidated apis.ConditionType = "consolidated"
```
Comment (thomaschaaf) [inline pkg/apis/v1beta1/nodeclaim_status.go:65]:
```suggestion
	Consolidated apis.ConditionType = "Consolidated"
```

**Comment by Luke-Smartnews** [comment]
Tested in our dev env, working without any issues. Will do more load tests.

**Comment by almson** [comment]
Can you write a description of how you've implemented this? This seems to drastically change how "WhenUnderutilized" works. Previously, it took effect when the cluster as a whole was underutilized, but now it seems to just mean that the node is underutilized?

**Comment by Bryce-Soghigian** [comment]
> Can you write a description of how you've implemented this? This seems to drastically change how "WhenUnderutilized" works. 

Further than a description it should probably be first proposed as an RFC to get reviewed by the community, and to talk about alternative approaches.

**Comment by Luke-Smartnews** [comment]
will do after I confirm it works on a large scale.

**Comment by Luke-Smartnews** [comment]
## Test Report

- nodepool settings
   ```yaml
    spec:
      disruption:
        budgets:
        - nodes: 10%
        consolidateAfter: 5m
        consolidationPolicy: WhenUnderutilized
        expireAfter: Never
        utilizationThreshold: 65
   ```

- confirmed `Underutilized` added to target nodes

  ```yaml
    conditions:
    - lastTransitionTime: "2024-02-23T11:24:05Z"
      status: "True"
      type: Initialized
    - lastTransitionTime: "2024-02-23T11:22:54Z"
      status: "True"
      type: Launched
    - lastTransitionTime: "2024-02-23T11:24:05Z"
      status: "True"
      type: Ready
    - lastTransitionTime: "2024-02-23T11:23:40Z"
      status: "True"
      type: Registered
    - lastTransitionTime: "2024-02-23T11:31:53Z"
      severity: Warning
      status: "True"
      type: Underutilized
  ```
- pending pods scheduled in 5m
  Screenshot 2024-02-23 at 20 47 21

- nodes are removed 5m later after pods deletion
   Screenshot 2024-02-23 at 20 47 47

**Comment by njtran** [comment]
Hey @Luke-Smartnews, this is a core difference in how Karpenter considers underutilization. We've intentionally not surfaced a utilization threshold due to its edge cases and how it ends up driving overall lower utilization. This would be a huge change to our disruption logic, and would definitely require opening up a design + RFC. I'm not sure we would be  accepting this feature, but I would love to hear about the use-cases you're trying to solve here.

**Comment by barryrobison** [comment]
> Hey @Luke-Smartnews, this is a core difference in how Karpenter considers underutilization. We've intentionally not surfaced a utilization threshold due to its edge cases and how it ends up driving overall lower utilization. This would be a huge change to our disruption logic, and would definitely require opening up a design + RFC. I'm not sure we would be accepting this feature (at least in its current state without a design), but I would love to hear about the use-cases you're trying to solve here.

@njtran is the current utilisation logic documented somewhere? From my brief reading of the source, it looks like karpenter does a "fake" scheduling run, and if the nodes pods can be scheduled elsewhere it consolidates the node?

Our use case is flink jobs performing data transfer. Those pods are bursty in nature, and we want to insure those pods run for at least 10 minutes so they can reach a checkpoint before getting rescheduled. The current system is reclaiming nodes in under 90s, so they barely have time to get scheduled before karpenter pulls the rug.

Thank you!

**Comment by njtran** [comment]
@barryrobison I think you want this https://github.com/kubernetes-sigs/karpenter/issues/752, you're correct in that's how Karpenter considers Consolidation

**Comment by Luke-Smartnews** [comment]
@njtran 
my change only take effect when the nodepool is configured whit `UtilizationThreshold`, it doesn't change the existing behavior at all.

https://github.com/kubernetes-sigs/karpenter/pull/992/files#diff-c11b5a9240bbfac3dd14fdfff84e098ade0f3bf0a5fc63673de41ff710d3d308R105

**Comment by njtran** [comment]
Hey @Luke-Smartnews, even if this is just a change to the candidacy, we'd need to review a design/RFC before continuing on this. Would you be able to open a design PR first?

**Comment by njtran** [comment]
Will close this until there's an RFC for this. Please feel free to re-open when you cut the RFC!

**Pull request closed without merging**